File: todo-backend/src/main/java/com/example/todo/service/TodoService.java
package com.example.todo.service;

import com.example.todo.model.Todo;
import com.example.todo.repository.TodoRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public class TodoService {

    private final TodoRepository todoRepository;

    @Autowired
    public TodoService(TodoRepository todoRepository) {
        this.todoRepository = todoRepository;
    }

    // Get all todos or filter by completed status
    public List<Todo> getTodosByStatus(Boolean completed) {
        if (completed != null) {
            return todoRepository.findByCompleted(completed);
        }
        return todoRepository.findAll();
    }

    // Get a single todo by ID
    public Optional<Todo> getTodoById(Long id) {
        return todoRepository.findById(id);
    }

    // Create a new todo
    public Todo createTodo(Todo todo) {
        todo.setCompleted(false);
        return todoRepository.save(todo);
    }

    // Update an existing todo
    public Todo updateTodo(Long id, Todo todoDetails) {
        Todo todo = todoRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Todo not found with id: " + id));

        todo.setTitle(todoDetails.getTitle());
        todo.setDescription(todoDetails.getDescription());
        todo.setCompleted(todoDetails.getCompleted());

        return todoRepository.save(todo);
    }

    // Toggle the completion status
    public Todo toggleTodoStatus(Long id) {
        Todo todo = todoRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Todo not found with id: " + id));

        todo.setCompleted(!todo.getCompleted());
        return todoRepository.save(todo);
    }

    // Delete a todo
    public void deleteTodo(Long id) {
        Todo todo = todoRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Todo not found with id: " + id));
        todoRepository.delete(todo);
    }
}

File: todo-frontend/src/services/todoService.js
import axios from 'axios';

const API_BASE_URL = 'http://localhost:8080/api/todos';

class TodoService {
    // Get all todos
    getAllTodos(completed = null) {
        const url = completed !== null
            ? `${API_BASE_URL}?completed=${completed}`
            : API_BASE_URL;
        return axios.get(url);
    }

    // Get todo by ID
    getTodoById(id) {
        return axios.get(`${API_BASE_URL}/${id}`);
    }

    // Create new todo
    createTodo(todo) {
        return axios.post(API_BASE_URL, todo);
    }

    // Update todo
    updateTodo(id, todo) {
        return axios.put(`${API_BASE_URL}/${id}`, todo);
    }

    // Toggle todo completion
    toggleTodoStatus(id) {
        return axios.patch(`${API_BASE_URL}/${id}/toggle`);
    }

    // Delete todo
    deleteTodo(id) {
        return axios.delete(`${API_BASE_URL}/${id}`);
    }
}

export default new TodoService();